### PR TITLE
feat(openinference-core): support metadata attributes on context

### LINKED ISSUE #764: [js][openinference-core] inheritable metadata attributes support

### PR DESCRIPTION
**Pull request by Parker-Stafford**
resolves #764 

- adds support for metadata attributes on context

### CODE REVIEW

File: js/packages/openinference-core/src/types.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export type GenericFunction = (...args: any[]) => any;
```
Comment (Parker-Stafford) [inline js/packages/openinference-core/src/types.ts:2]: we can remove these and the withSafety util below from other packages if we want to export all of this stuff out of this package

File: js/packages/openinference-core/test/trace/contextAttributes.test.ts [hunk 2]
```diff
@@ -2,9 +2,11 @@ import { context, ContextManager } from "@opentelemetry/api";
 import { AsyncHooksContextManager } from "@opentelemetry/context-async-hooks";
 
 import {
+  METADATA,
```
Comment (Parker-Stafford) [inline js/packages/openinference-core/test/trace/contextAttributes.test.ts:5]: going to pr main to add support for these two tings then rebase / remove from feature branch

File: js/packages/openinference-semantic-conventions/src/trace/SemanticConventions.ts [hunk 3]
```diff
@@ -337,6 +337,11 @@ export const SESSION_ID =
 export const USER_ID =
   `${SemanticAttributePrefixes.user}.${UserAttributePostfixes.id}` as const;
 
+/**
+ * Metadata for a span, used to store user-defined key-value pairs
+ */
+export const METADATA = "metadata" as const;
```
Comment (Parker-Stafford) [inline js/packages/openinference-semantic-conventions/src/trace/SemanticConventions.ts:343]: will remove after pushign to main

File: js/packages/openinference-core/src/typeUtils.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import { AttributeValue } from "@opentelemetry/api";
+
```
Comment (mikeldking) [inline js/packages/openinference-core/src/typeUtils.ts:2]: can you add jsDocs here?

File: js/packages/openinference-core/src/types.ts (hunk 1 above)
Comment (mikeldking) [reply js/packages/openinference-core/src/types.ts:2]: let's move this into the /utils directory

File: js/packages/openinference-core/test/trace/contextAttributes.test.ts [hunk 5]
```diff
@@ -130,3 +169,58 @@ describe("context.with multiple attributes", () => {
...
+        const attributes = getAttributesFromContext(context.active());
+        expect(attributes).toStrictEqual({
+          [PROMPT_TEMPLATE_TEMPLATE]: "hello {name}",
+          [PROMPT_TEMPLATE_VARIABLES]: JSON.stringify(variables),
+          [PROMPT_TEMPLATE_VERSION]: "V1.0",
+          [SESSION_ID]: "session-id",
+          [METADATA]: JSON.stringify({ key: "value" }),
+        });
+      },
+    );
```
Comment (mikeldking) [inline js/packages/openinference-core/test/trace/contextAttributes.test.ts:209]: Can you add to the tests that these attributes get un-set?

**Review by mikeldking — APPROVED** [review]
Nice

File: js/packages/openinference-core/test/trace/contextAttributes.test.ts (hunk 5 above)
Comment (Parker-Stafford) [reply js/packages/openinference-core/test/trace/contextAttributes.test.ts:209]: get un-set from the context or un-set from where? Right now getAttributes from context doesn't do any un-setting although we could, that wouldn't allow you to read the context attributes more than once though which seems wrong

**Pull request merged**